### ISSUE TITLE
@Before* in outer and @Nested classes interaction

### ISSUE THREAD

**Issue by Sam-Kruglov** [junit-team/junit5#1994]
I can't compose a single thought of what I can't get working, so I've prepared a test case in Kotlin that will describe what I am trying to do and the problems are highlighted with `fixme` comments. There is only one problem here that is repeated 4 times but I think it may be useful for you to look at the whole picture. Please, notice, that all classes here use `TestInstance.Lifecycle.PER_CLASS` strategy.
```kotlin
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.MethodOrderer
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Order
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.api.TestMethodOrder

class EntityRepo {
    fun save(entity: String): String {
        println("saving $entity")
        return entity
    }
}

class RequestSender {
    fun send(request: String) = println("sending $request")
}

//some annotations and stuff
//todo check up on @Nested https://github.com/spring-projects/spring-framework/issues/19930
abstract class AbstractIntegrationTest

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@TestMethodOrder(MethodOrderer.OrderAnnotation::class)
abstract class StatefulIT : AbstractIntegrationTest()

abstract class ClearsDbBeforeAllIT : StatefulIT() {
    @BeforeAll
    fun clearDbBeforeAll() {
        println("clear db")
    }
}

class StartEntityIT : StatefulIT() {

    val requestSender = RequestSender()
    val entityRepo = EntityRepo()

    @Nested
    inner class `given a STOPPED entity in DB` : ClearsDbBeforeAllIT() {

        lateinit var stoppedEntity: String

        @BeforeAll
        fun setUp() {
            stoppedEntity = entityRepo.save("stopped entity")
        }

        @Test
        fun `get entity -- returned with STOPPED status`() {
            requestSender.send("GET /api/entities/{id} for $stoppedEntity")
            //assert returned status is STOPPED
        }

        @Nested
        inner class `given STOPPED in Vendor` : ClearsDbBeforeAllIT() {

            @BeforeAll
            fun setUp1() {
                //fixme stopped entity is not saved
                println("making vendor api have STOPPED status and accept 'start'")
            }

            @Test
            fun `start -- success`() {
                requestSender.send("PUT /api/entities/{id}/start for $stoppedEntity")
            }
        }

        @Nested
        inner class `given DISABLED in Vendor` : ClearsDbBeforeAllIT() {

            @BeforeAll
            fun setUp() {
                //fixme stopped entity is not saved
                println("making vendor api have DISABLED status and reject 'start'")
            }

            @Order(1)
            @Test
            fun `start -- reject`() {
                requestSender.send("PUT /api/entities/{id}/start for $stoppedEntity")
            }

            @Order(2)
            @Test
            fun `entity in DB has synchronized with Vendor to have DISABLED status`() {
                requestSender.send("GET /api/entities/{id} for $stoppedEntity")
                //assert returned status is DISABLED
            }
        }
    }

    @Nested
    inner class `given a STARTED entity in DB` : ClearsDbBeforeAllIT() {

        lateinit var startedEntity: String

        @BeforeAll
        fun setUp() {
            startedEntity = entityRepo.save("started entity")
        }

        @Nested
        inner class `given STOPPED in Vendor` : ClearsDbBeforeAllIT() {

            @BeforeAll
            fun setUp() {
                //fixme started entity is not saved
                println("making vendor api have STOPPED status and accept 'start'")
            }

            @Test
            fun `start -- success`() {
                requestSender.send("PUT /api/entities/{id}/start for $startedEntity")
            }
        }

        @Nested
        inner class `given STARTED in Vendor` : ClearsDbBeforeAllIT() {

            @BeforeAll
            fun setUp() {
                //fixme started entity is not saved
                println("making vendor api have STARTED status and accept 'start'")
            }

            @Test
            fun success() {
                requestSender.send("PUT /api/entities/{id}/start for $startedEntity")
            }
        }
    }
}
```

Jupiter version: 5.5.1

this may or may not be connected with #1991

---
Here is the output:

```
clear db
saving stopped entity

sending GET /api/entities/{id} for stopped entity

clear db
making vendor api have STOPPED status and accept 'start'

sending PUT /api/entities/{id}/start for stopped entity

clear db
making vendor api have DISABLED status and reject 'start'

sending PUT /api/entities/{id}/start for stopped entity


sending GET /api/entities/{id} for stopped entity

clear db
saving started entity
clear db
making vendor api have STOPPED status and accept 'start'

sending PUT /api/entities/{id}/start for started entity

clear db
making vendor api have STARTED status and accept 'start'

sending PUT /api/entities/{id}/start for started entity

```

**Comment by sbrannen** [comment]
Tentatively slated for the 5.6 backlog for _team discussion_ and further investigation

**Comment by marcphilipp** [comment]
@Sam-Kruglov I've read through the sample code you've provided but I'm still unsure what you're trying to achieve. I looks like you'd only want to use `ClearsDbBeforeAllIT` on the outermost `@Nested` class otherwise you're clearing the database before each inner class. Instead of clearing, could each `@Nested` class delete the entities it saved in an `@AfterAll` method?

**Comment by Sam-Kruglov** [comment]
@marcphilipp I want to clear db before each nested test because they are independent but modify db state. However, I still want to share the common set up part like I do in my unit tests.

```
class "given X"
  @nested

**Comment by marcphilipp** [comment]
This works as designed. `@BeforeEach` steps in nested classes are supposed to build on each other. If you don't want that why use nested tests at all?

```
class "given X"
  tests 1
class "given Y"
  tests 2
class "given Z"
  tests 3
```

**Comment by Sam-Kruglov** [comment]
Because I want to write tests under "given xyz" and under "given by" also so I have a nice decision tree

**Comment by marcphilipp** [comment]
If you have setup logic that you only need for tests 1 but not tests 2/3, you could wrap it in another nested class.

I'm closing this issue as _works-as-designed_.

**Comment by Sam-Kruglov** [comment]
Sorry, I typed `by` instead of `xy` in my last comment. 

Could you give an example, it is not very clear to me. 

How do I make that tree work, for example?
```
given X
  test 1
  given Y
    test 2
    test 3
    given Z
      test 4
```

**Comment by sormuras** [comment]
Like this?

https://junit.org/junit5/docs/current/user-guide/#writing-tests-nested

**Comment by Sam-Kruglov** [comment]
If you look at my original example:

```
    @Nested
    inner class `given a STOPPED entity in DB` : ClearsDbBeforeAllIT() {

        lateinit var stoppedEntity: String

        @BeforeAll
        fun setUp() {
            stoppedEntity = entityRepo.save("stopped entity")
        }

        @Test
        fun `get entity -- returned with STOPPED status`() {
            requestSender.send("GET /api/entities/{id} for $stoppedEntity")
            //assert returned status is STOPPED
        }

        @Nested
        inner class `given STOPPED in Vendor` : ClearsDbBeforeAllIT() {

            @BeforeAll
            fun setUp1() {
                //fixme stopped entity is not saved
                println("making vendor api have STOPPED status and accept 'start'")
            }

            @Test
            fun `start -- success`() {
                requestSender.send("PUT /api/entities/{id}/start for $stoppedEntity")
            }
        }
    }
```

We get this output 
```
clear db
saving stopped entity

sending GET /api/entities/{id} for stopped entity

clear db // !!! this must not be here because the next PUT we send will not find the entity
making vendor api have STOPPED status and accept 'start'

sending PUT /api/entities/{id}/start for stopped entity
```

So, what @sormuras suggested does not help with this problem.

**Comment by marcphilipp** [comment]
@Sam-Kruglov If you use `@BeforeEach` instead of `@BeforeAll` it would work. If you have to keep `@BeforeAll`, you'd need a common super class of both `@Nested` classes and move the `@BeforeAll` there.

**Comment by Sam-Kruglov** [comment]
It seems it would quickly become pretty messy. I would really like to have a better way of writing stateful (test instance per class) tree-like (deeply @Nested) integration tests. Currently, I can only do that nicely with unit tests because they are stateless (test instance per method)

**Comment by marcphilipp** [comment]
You can use `@BeforeEach` with "stateful" tests, can't you?

**Comment by Sam-Kruglov** [comment]
If I set up the test state before each test then it's not stateful anymore. What do you mean?

**Comment by marcphilipp** [comment]
Why don't you extend `ClearsDbBeforeAllIT` only on the outermost test class?

**Comment by Sam-Kruglov** [comment]
Because I want to recreate the state completely before entering the inner class:
```
setup: { clear db; save entity }
def: { some generic tests that mutate state }
nested 1: setup: { nested1_setup } def: { more tests that mutate state }
  nested 11: setup: { nested11_setup } def: { specific tests that mutate state }
  nested 12: setup: { nested12_setup } def: { specific tests that mutate state }
  nested 13: setup: { nested13_setup } def: { specific tests that mutate state }
nested 2: ...
```
So, before going into any class I want to perform the setup all the way to from the top level class like that:
```
clear db
save entity
outer generic tests

clear db
save entity
nested1_setup
nested1_tests

clear db
save entity
nested1_setup
nested11_setup
nested11_tests

clear db
save entity
nested1_setup
nested12_setup
nested12_tests

clear db
save entity
nested1_setup
nested13_setup
nested13_tests

clear db
save entity
nested2_setup
...
```

**Comment by Sam-Kruglov** [comment]
@marcphilipp let me know if you have any more questions

**Comment by marcphilipp** [comment]
That's exactly how it would work if you used `@BeforeEach`. If you need it to work with `@BeforeAll` you could call `nested1_setup` from `nested12_setup` and so forth.

**Comment by Sam-Kruglov** [comment]
If I want to insert and element in one test and then search it in another test, I won't find it if I clear the database before each test, it doesn't make sense.
I guess manually call outer class' setup in your setup is an option, but quite error-prone, do you think it would be possible to put the automated approach on the backlog?

**Comment by marcphilipp** [comment]
I think the current behavior is correct so I don't really see that we should change it when `@TestInstance(PER_CLASS)` is used.

@junit-team/junit-lambda WDYT?

**Comment by Sam-Kruglov** [comment]
Maybe I just don't understand how I am supposed to use Junit 5 for integration tests? 
My assumption that the best way, in general, is to use `@Nested` everywhere so that there will be a nice decision tree. For integration tests, I have to use `@TestInstance(PER_CLASS)` and `@TestMethodOrder`. If I don't use these 2 annotations, and instead use `@BeforeEach` in every step, then I think it is too much of an overhead for integration tests: I have to clear the whole database, reinstantiate some queues, etc for every test. And I try to keep my tests to have a single assertion only which makes this repetition too high. I

Basically, if I don't do it the way I described, I think I'd have to not use `@Nested` for integration tests at all and simply have each test have a huge name and handle all the setup and all the assertions. There will be a lot of copy paster.

If I do use that `@Nested` approach, I have to call the outer class' `@BeforeAll` method manually which makes me think I do something wrong.
